pkg/net/bitswap.go

```diff
@@ -187,58 +193,90 @@ func (c BitswapClient) SpadeTraversal(parent context.Context,
+			}
+
+			// if not at bottom of the tree, keep going down the links until we reach it or hit a dead end
+			links, err := FindLinks(parent, blk)
+			if err != nil {
+				return task.NewErrorRetrievalResultWithErrorResolution(task.CannotDecodeLinks, err), nil
+			}
+
+			logger.Debugf("cid %s has %d links\n", cidToRetrieve.String(), len(links))
+
+			nextIndexes = append(nextIndexes, links...)
 		}
 
-		if i == maxTraverseDepth {
+		if len(nextIndexes) == 0 {
 			var size = int64(len(blk.RawData()))
 			elapsed := time.Since(startTime)
 			logger.With("size", size).With("elapsed", elapsed).Info("Retrieved block")
 
-			// we've reached the requested depth of the tree
 			return task.NewSuccessfulRetrievalResult(elapsed, size, elapsed), nil
 		}
 
-		// if not at bottom of the tree, keep going down the links until we reach it or hit a dead end
-		links, err := FindLinks(parent, blk)
-		if err != nil {
-			return task.NewErrorRetrievalResultWithErrorResolution(task.CannotDecodeLinks, err), nil
-		}
+		// Clear out cids list to prevent resizing array
+		cidsToRetrieve = cidsToRetrieve[:0]
 
-		logger.Debugf("cid %s has %d links\n", cidToRetrieve.String(), len(links))
-
-		if len(links) == 0 {
-			var size = int64(len(blk.RawData()))
-			elapsed := time.Since(startTime)
-			logger.With("size", size).With("elapsed", elapsed).Info("Retrieved block")
+		// Shuffle a slice so we can find the next cids to challenge
+		rand.Shuffle(len(nextIndexes), func(i, j int) {
+			nextIndexes[i], nextIndexes[j] = nextIndexes[j], nextIndexes[i]
+		})
 
-			return task.NewSuccessfulRetrievalResult(elapsed, size, elapsed), nil
+		for j := 0; j < int(math.Min(float64(numChallenge), float64(len(nextIndexes)))); j++ {
+			cid, err := cid.Parse(nextIndexes[j].String())
+			if err != nil {
+				return task.NewErrorRetrievalResultWithErrorResolution(task.CIDCodecNotSupported, err), nil
+			}
+			cidsToRetrieve = append(cidsToRetrieve, cid)
 		}
 
-		// randomly pick a link to go down
-		//nolint:all we don't need crypto secured random numbers
-		nextIndex := rand.Intn(len(links))
+		// Clear out nextIndexes list to help prevent resizing array
+		nextIndexes = nextIndexes[:0]
 
-		cidToRetrieve, err = cid.Parse(links[nextIndex].String())
-		if err != nil {
-			return task.NewErrorRetrievalResultWithErrorResolution(task.CIDCodecNotSupported, err), nil
-		}
+		// To the next layer of the tree
+		i++
+	}
+}
+
+func initializeCidsToRetrieve(numChallenge int, startingCid cid.Cid) []cid.Cid {
+	cidsToRetrieve := make([]cid.Cid, 0, numChallenge)
+	cidsToRetrieve = append(cidsToRetrieve, startingCid)
+	return cidsToRetrieve
+}
 
-		i++ // To the next layer of the tree
+func validateAndGetNumberOfChallenges(logger *zap.SugaredLogger) int {
+	var numChallenge = env.GetInt(env.NumberOfChallenges, 0)
+	if numChallenge <= 0 || numChallenge > 10 {
+		numChallenge = 1
+		logger.Infof("NumberOfChallenges is <= 0 or NumberOfChallenges > 10. Using 1 challenge per level")
```

Comment:
I dont think we need to constrain it to 10 here since it won't result in exponential number of queries. There could be a few hundred of CIDs per level that they wanna check

Reply:
do we want exponential number of queries per level? I dont have a feeling how deep these will go but we could be potentially using a lot of resources,